typing/ctype.ml

```diff
@@ -2652,18 +2671,8 @@ let rec unify (env:Env.t ref) t1 t2 =
         update_level !env t1.level t2;
         update_scope t1.scope t2;
         link_type t1 t2
-    | (Tconstr (p1, [], _), Tconstr (p2, [], _))
-      when Env.has_local_constraints !env
-      && is_newtype !env p1 && is_newtype !env p2 ->
-        (* Do not use local constraints more than necessary *)
-        begin try
-          if find_expansion_scope !env p1 > find_expansion_scope !env p2 then
-            unify env t1 (try_expand_once !env t2)
-          else
-            unify env (try_expand_once !env t1) t2
-        with Cannot_expand ->
-          unify2 env t1 t2
-        end
+    | (Tconstr _, Tconstr _) when Env.has_local_constraints !env ->
+        (try unify_public !env t1 t2  with Cannot_expand -> unify2 env t1 t2)
```

Comment:
Rather than raising `Cannot_expand`, would it be better to just call straight into `unify2` from `unify_public`? It seems like that might save us from redoing some expansion work at the start of `unify2`.

Reply:
As indicated above, `unify2` should be called with the non-expanded version of the type.
Note that thanks to memoization, redoing expansions just amounts to looking them up , without any allocation.
This small cost could be avoided by having `unify2` receive both the original and partly expanded forms, like `unify3`. Since we only do extra work when expanding non-parametric type abbreviations, I'm not sure it is worth adding extra complexity to the code to optimize that case.

Reply:
> having unify2 receive both the original and partly expanded forms

Yes, that is what I had in mind. I think it is a tiny change and if it would allow us to remove the `is_public_type` conditions then I think the code would overall be simpler.

> Since we only do extra work when expanding non-parametric type abbreviations

I don't see why the work is only for non-parametric type abbreviations. It seems to be for all type abbreviations. I think that is fine, especially if we reuse the work by adding the extra parameters to `unify2`.